packages/react/src/components/MultiSelect/__tests__/MultiSelect-test.js

```diff
@@ -528,4 +530,84 @@ describe('MultiSelect', () => {
       );
     });
   });
+
+  describe('Controlled', () => {
+    const ControlledMultiselect = () => {
+      const items = generateItems(4, generateGenericItem);
+      const [selectedItems, setSelectedItems] = useState([]);
+
+      const onSelectionChanged = (value) => {
+        setSelectedItems(value);
+      };
+      return (
+        <>
+          <MultiSelect
+            id="test"
+            titleText="Multiselect title"
+            label="test-label"
+            items={items}
+            selectedItems={selectedItems}
+            onChange={(data) => onSelectionChanged(data.selectedItems)}
+            selectionFeedback="top-after-reopen"
+          />
+          <br />
+          <ButtonSet>
+            <Button
+              id="all"
+              onClick={() =>
+                setSelectedItems(items.filter((item) => !item.disabled))
+              }>
+              Select all
+            </Button>
+            <Button
+              id="clear"
+              kind="secondary"
+              onClick={() => setSelectedItems([])}>
+              Clear
+            </Button>
+          </ButtonSet>
+        </>
+      );
+    };
+
+    it('should initially render controlled multiselect with a given label', () => {
+      const label = 'test-label';
+      const { container } = render(<ControlledMultiselect />);
+
+      // eslint-disable-next-line testing-library/prefer-screen-queries
+      const labelNode = getByText(container, label);
+      expect(isElementVisible(labelNode)).toBe(true);
+
+      expect(
+        // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
+        container.querySelector(
+          '[aria-expanded="true"][aria-haspopup="listbox"]'
+        )
+      ).toBeNull();
+    });
+    it('should allow the items to be controlled from external state', async () => {
```

Comment:
The only negative I see with this approach is that the test is inherently impure. It primarily tests `ControlledComponent`, and only by proxy the `MultiSelect`'s `selectedItems` and `onChange`. If either `ControlledComponent` or `MultiSelect` were to change, there's a (small) possibility the "fuzziness" of this test would make it still pass when it shouldn't or vice versa. 

Check out [this article](https://moxio.com/blog/4-methods-for-testing-controlled-react-components/) - it does a better job explaining some approaches to this. The test you have now is most of "Route 4", just without the spy, which would be a great thing to add to this. 

Another option would be "Route 3" using `rerender` and an `onChange` mock, which also seems like a decent option. Maybe we have both! Here's a rough idea of "Route 3":



```js
it('should support controlled component functionality with selectedItems and onChange', async () => {
  const onChange = jest.fn();
  const { rerender } = render(<MultiSelect items={items} selectedItems={[items[0]]} onChange={onChange} />);

  // The selected items should match what's passed into selectedItems
  expect(screen.getAllByRole('option')[0]).toHaveAttribute('aria-selected', 'true');

  // onChange should fire for interactions
  await userEvent.click(screen.getAllByRole('option')[1])
  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange).toHaveBeenCalledWith([items[1]]);

  // If the onChange event data is not used to update selectedItems, the selection should remain as it was before
  expect(screen.getAllByRole('option')[0]).toHaveAttribute('aria-selected', 'true');
  expect(screen.getAllByRole('option')[1]).toHaveAttribute('aria-selected', 'false');

  // Force a re-render with updated selectedItems matching the initial selection and the onChange event data
  render(<MultiSelect items={items} selectedItems={[items[0], items[1]]} onChange={onChange} />);

  // Now both should be selected
  expect(screen.getAllByRole('option')[0]).toHaveAttribute('aria-selected', 'true');
  expect(screen.getAllByRole('option')[1]).toHaveAttribute('aria-selected', 'true');
});
```

Reply:
Added a test suite looking at this example, please let me know of any changes. I have kept the "route-4" type of test also as an example for us on how it could work from the test side, so that users can refer to it and write their tests for their implementation.